# ДЗ

#### 1) Попытаться ответить на вопросы/выдвинуть гипотезы

- Чего хочет менеджер от рекомендательной системы? (рост показателя X на Y% за Z недель)

###### Думаю менеджер хочет увеличить прибыль. В таком случае, наилучшим показателем будет увеличение прибыли на каждый вложенный рубль в разработку рекомендательной системы

- По-хорошему надо бы предварительно посчитать потенциальный эффект от рекоммендательной системы (Оценки эффектов у менеджера и у вас могут сильно не совпадать: как правило, вы знаете про данные больше)

###### Нужно задать цель. Например увеличить прибыль на 10%, ну и дополнительные условия при которых принимать решения об эффективности/неэффективности системы. Как пользователь могу сказать что на сегодня рекомендации в большенстве своём носят реактивный характер и чаще раздрожают чем помогают. Например, я уже купил что-то а мне продолжает всплывать контекстная реклама. Или хотел сделать жене сюрприз на ДР а Google меня спалил показав в рекомендациях что я искал.

- А у нас вообще есть e-mail-ы пользователей? Для скольки %? Не устарели ли они?

###### Это важно. Есть смысл обучить систему так, чтобы с какой-то верояностью выделялись пользователи которых раздрожает спам и не присылать им сообщения. Сейчас шлют всем подряд. Мне один из банков предлагает воспользоваться их услугами когда сами заблокировали счета. Не знаю как в других странах, но у нас это пока любительство. Но работать в этом направлении конечно нужно.

- Будем ли использовать СМС и push-уведомления в приложении? Может, будем печатать рекомендации на чеке после оплаты на кассе?

###### См. предыдущий ответ по E-mail.

- Как будет выглядеть e-mail? (решаем задачу топ-10 рекомендаций или ранжирования? И топ-10 ли?)

###### С 10 товарами e-mail будет выглядеть никак, т.к. сразу отправится в СПАМ и сообщение никто не увидит. 3 -максимум, но это моё мнение.

- Какие товары должны быть в e-mail? Есть ли какие-то ограничения (только акции и т п)?

###### Должно быть что-то специфическое для этого пользователя.

- Сколько денег мы готовы потратить на привлечение 1 юзера? CAC - Customer Aquisition Cost. Обычно CAC = расходы на коммуникацию + расходы на скидки

###### Принципиальный вопрос, значение которого переоценить очень сложно.

- Cколько мы хотим зарабатывать с одного привлеченного юзера?

###### Тоже важно, т.к. нужно знать цель.

---
- А точно нужно сортировать по вероятности?

###### Да, чтобы не поподались фары с платьями.

- Какую метрику использовать?

###### На мой взгляд это однозначно Precision. Лучше не послать кому-то сообщение, чем спугнуть спамом хорошего клиента

- Сколько раз в неделю отпрпавляем рассылку?

###### Максимум 1 раз в неделю, а то и пару раз в месяц.

- В какое время отправляем рассылку?

###### У всех время разное, особенно в такой большой стране как Россия. Так же зависит от товара, если это товары для отдых, то пятница вечер это наилучший выбор ну и т.д.

- Будем отправлять одному юзеру много раз наши рекоммендации. Как добиться того, чтобы они хоть немного отличались?

###### Постоянно оптимизировать алгоритм. Полностью исключить повторы одних и тех же рекомендаций, иначе это выглядит смешно в лучшем случае и назойливо в большинстве случаев.

- Нужно ли, чтобы в одной рассылке были *разные* товары? Как определить, что товары *разные*? Как добиться того, чтобы они были разными?

###### Не думаю что это необходимо. Добитсяможно брать ближайших соседей по группе товаров или аналоги других производителей, с технической точки зрения не вижу проблем берем аналог или соседа, сравниваем с тем что уже отправлялось, если ответ да меняем на следующего.

- И многое другое:)

###### И т.д. :)

- Хотим повысить выручку минимум на 6% за 4 месяца. Будем повышать за счет роста Retention минимум на  3% и среднего чека минимум на 3%

###### Похвальное желание. Так держать :)

- Топ-5 товаров, а не топ-10 (В e-mail 10 выглядят не красиво, в push и на чек больше 5 не влезает)

###### Сделать максимум 3. Не благодарите :)

- Рассылаем в e-mail (5% клиентов) и push-уведомлении (20% клиентов), печатаем на чеке (все оффлайн клиенты)

###### Молодцы! :)

- **3 товара с акцией** (Как это учесть? А если на товар была акция 10%, а потом 50%, что будет стоять в user-item матрице?)

###### Не совсем понял вопрос, записываем максимальную скидку чтобы привлечь внимание.

- **1 новый товар** (юзер никогда не покупал. Просто фильтруем аутпут ALS? А если у таких товаров очень маленькая вероятность покупки? Может, использовать другую логику/модель?) 

###### Думаю есть смысл найти похожего пользователя (параметры всегда найдутся: регион по IP, просмотренные товары, время регистрации, про пол и возвраст и говорить нечего, но они не всегда бывают). Второй вариант, самые популярные товары.

- **1 товар для роста среднего чека** (товары минимум дороже чем обычно покупает юзер. Как это измерить? На сколько дороже?)

###### Думаю 10-15% дороже не больше. Просто предлагаем более дорогой популярный аналог товара/услуги котрорый купил/просмотрел пользователь и не мучаем клавиатуру.


## Практика:

##### 2) Поэкспериментировать с ALS (grid-search)

In [117]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# Для работы с матрицами
from scipy.sparse import csr_matrix

# Матричная факторизация
from implicit.als import AlternatingLeastSquares
from implicit.nearest_neighbours import bm25_weight, tfidf_weight

# Функции из 1-ого вебинара
import os, sys

module_path = os.path.abspath(os.path.join(os.pardir))
if module_path not in sys.path:
    sys.path.append(module_path)
    
from metrics import precision_at_k, recall_at_k

import pickle

from scipy.spatial.distance import cdist

from sklearn.decomposition import PCA
from sklearn.manifold import TSNE

import seaborn as sns

In [2]:
data = pd.read_csv('data/transaction_data.csv')

data.columns = [col.lower() for col in data.columns]
data.rename(columns={'household_key': 'user_id',
                    'product_id': 'item_id'},
           inplace=True)


test_size_weeks = 3

data_train = data[data['week_no'] < data['week_no'].max() - test_size_weeks]
data_test = data[data['week_no'] >= data['week_no'].max() - test_size_weeks]

data_train.head(10)

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
0     2375  26984851472    1  1004906         1         1.39       364   
1     2375  26984851472    1  1033142         1         0.82       364   
2     2375  26984851472    1  1036325         1         0.99       364   
3     2375  26984851472    1  1082185         1         1.21       364   
4     2375  26984851472    1  8160430         1         1.50       364   
5     2375  26984851516    1   826249         2         1.98       364   
6     2375  26984851516    1  1043142         1         1.57       364   
7     2375  26984851516    1  1085983         1         2.99       364   
8     2375  26984851516    1  1102651         1         1.89       364   
9     2375  26984851516    1  6423775         1         2.00       364   

   retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0        -0.60        1631        1          0.0                0.0  
1         0.00        1631        1          0.0                0.0  
2        -0.30        1631        1          0.0                0.0  
3         0.00        1631        1          0.0                0.0  
4        -0.39        1631        1          0.0                0.0  
5        -0.60        1642        1          0.0                0.0  
6        -0.68        1642        1          0.0                0.0  
7        -0.40        1642        1          0.0                0.0  
8         0.00        1642        1          0.0                0.0  
9        -0.79        1642        1          0.0                0.0

In [3]:
item_features = pd.read_csv('data/product.csv')
item_features.columns = [col.lower() for col in item_features.columns]
item_features.rename(columns={'product_id': 'item_id'}, inplace=True)

item_features.head(2)

item_id  manufacturer    department     brand            commodity_desc  \
0    25671             2       GROCERY  National                  FRZN ICE   
1    26081             2  MISC. TRANS.  National  NO COMMODITY DESCRIPTION   

            sub_commodity_desc curr_size_of_product  
0          ICE - CRUSHED/CUBED                22 LB  
1  NO SUBCOMMODITY DESCRIPTION

In [4]:
item_features.department.unique()

array(['GROCERY', 'MISC. TRANS.', 'PASTRY', 'DRUG GM', 'MEAT-PCKGD',
       'SEAFOOD-PCKGD', 'PRODUCE', 'NUTRITION', 'DELI', 'COSMETICS',
       'MEAT', 'FLORAL', 'TRAVEL & LEISUR', 'SEAFOOD', 'MISC SALES TRAN',
       'SALAD BAR', 'KIOSK-GAS', 'ELECT &PLUMBING', 'GRO BAKERY',
       'GM MERCH EXP', 'FROZEN GROCERY', 'COUP/STR & MFG', 'SPIRITS',
       'GARDEN CENTER', 'TOYS', 'CHARITABLE CONT', 'RESTAURANT', 'RX',
       'PROD-WHS SALES', 'MEAT-WHSE', 'DAIRY DELI', 'CHEF SHOPPE', 'HBC',
       'DELI/SNACK BAR', 'PORK', 'AUTOMOTIVE', 'VIDEO RENTAL', ' ',
       'CNTRL/STORE SUP', 'HOUSEWARES', 'POSTAL CENTER', 'PHOTO', 'VIDEO',
       'PHARMACY SUPPLY'], dtype=object)

In [5]:
result = data_test.groupby('user_id')['item_id'].unique().reset_index()
result.columns=['user_id', 'actual']
result.head(2)

user_id                                             actual
0        1  [879517, 934369, 1115576, 1124029, 5572301, 65...
1        3  [823704, 834117, 840244, 913785, 917816, 93870...

In [6]:
popularity = data_train.groupby('item_id')['quantity'].sum().reset_index()
popularity.rename(columns={'quantity': 'n_sold'}, inplace=True)

top_5000 = popularity.sort_values('n_sold', ascending=False).head(5000).item_id.tolist()

In [7]:
data_train.head(5)

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
0     2375  26984851472    1  1004906         1         1.39       364   
1     2375  26984851472    1  1033142         1         0.82       364   
2     2375  26984851472    1  1036325         1         0.99       364   
3     2375  26984851472    1  1082185         1         1.21       364   
4     2375  26984851472    1  8160430         1         1.50       364   

   retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0        -0.60        1631        1          0.0                0.0  
1         0.00        1631        1          0.0                0.0  
2        -0.30        1631        1          0.0                0.0  
3         0.00        1631        1          0.0                0.0  
4        -0.39        1631        1          0.0                0.0

In [8]:
# Заведем фиктивный item_id

data_train.loc[~data_train['item_id'].isin(top_5000), 'item_id'] = 999_999

user_item_matrix = pd.pivot_table(data_train, 
                                  index='user_id', columns='item_id', 
                                  values='quantity', # Можно пробоват ьдругие варианты
                                  aggfunc='count', 
                                  fill_value=0
                                 )

user_item_matrix = user_item_matrix.astype(float) # необходимый тип матрицы для implicit

# переведем в формат saprse matrix
sparse_user_item = csr_matrix(user_item_matrix)

user_item_matrix.head(3)

C:\Users\AVB\anaconda3\lib\site-packages\pandas\core\indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


item_id  202291    397896    420647    480014    545926    707683    731106    \
user_id                                                                         
1             0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2             0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3             0.0       0.0       0.0       0.0       0.0       0.0       0.0   

item_id  818980    819063    819227    ...  15926885  15926886  15926887  \
user_id                                ...                                 
1             0.0       0.0       0.0  ...       0.0       0.0       0.0   
2             0.0       0.0       0.0  ...       0.0       0.0       0.0   
3             0.0       0.0       0.0  ...       0.0       0.0       0.0   

item_id  15926927  15927033  15927403  15927661  15927850  16809471  17105257  
user_id                                                                        
1             2.0       0.0       0.0       0.0       0.0       0.0       0.0  
2             1.0       0.0       0.0       0.0       0.0       0.0       0.0  
3             0.0       0.0       0.0       0.0       0.0       0.0       0.0  

[3 rows x 5001 columns]

In [9]:
userids = user_item_matrix.index.values
itemids = user_item_matrix.columns.values

matrix_userids = np.arange(len(userids))
matrix_itemids = np.arange(len(itemids))

id_to_itemid = dict(zip(matrix_itemids, itemids))
id_to_userid = dict(zip(matrix_userids, userids))

itemid_to_id = dict(zip(itemids, matrix_itemids))
userid_to_id = dict(zip(userids, matrix_userids))

# ALS

In [10]:
%%time

model = AlternatingLeastSquares(factors=64, 
                                regularization=0.05,
                                iterations=15, 
                                calculate_training_loss=True, 
                                num_threads=10,
                                use_gpu=False)

model.fit(csr_matrix(user_item_matrix).T.tocsr(),  # На вход item-user matrix
          show_progress=True)



Wall time: 3.14 s


In [54]:
recs = model.recommend(userid=userid_to_id[2],  # userid - id от 0 до N
                        user_items=csr_matrix(user_item_matrix).tocsr(),   # на вход user-item matrix
                        N=5, # кол-во рекомендаций 
                        filter_already_liked_items=False, 
                        filter_items=None, 
                        recalculate_user=True)

In [55]:
[id_to_itemid[rec[0]] for rec in recs]

[999999, 1082185, 1133018, 5569230, 1106523]

In [47]:
def get_recommendations(user, model, N=5):
    res = [id_to_itemid[rec[0]] for rec in 
                    model.recommend(userid=userid_to_id[user], 
                                    user_items=sparse_user_item,   # на вход user-item matrix
                                    N=N, 
                                    filter_already_liked_items=False, 
                                    filter_items=None, 
                                    recalculate_user=True)]
    return res

In [99]:
%%time
    
result['als'] = result['user_id'].apply(lambda x: get_recommendations(x, model=model, N=5))

result.apply(lambda row: precision_at_k(row['als'], row['actual']), axis=1).mean()

Wall time: 19.8 s


0.16132596685082642

In [15]:
result.head(2)

user_id                                             actual  \
0        1  [879517, 934369, 1115576, 1124029, 5572301, 65...   
1        3  [823704, 834117, 840244, 913785, 917816, 93870...   

                                            als  
0    [832678, 1033142, 995242, 5569374, 979707]  
1  [1106523, 1133018, 910032, 1092026, 5569327]

In [16]:
model.item_factors.shape

(5001, 64)

In [17]:
model.user_factors.shape

(2500, 64)

In [18]:
fast_recs = model.user_factors @ model.item_factors.T 
fast_recs.shape

(2500, 5001)

In [19]:
fast_recs[0,:]

array([-0.00959435,  0.02364633, -0.02259211, ...,  0.05886692,
       -0.1647458 , -0.07703453], dtype=float32)

In [20]:
%%time
recommendations = model.recommend_all(N=5, 
                                      user_items=csr_matrix(user_item_matrix).tocsr(),
                                      filter_already_liked_items=True, 
                                      filter_items=None, 
                                      recalculate_user=True,
                                      show_progress=True,
                                      batch_size=500)
recommendations


Wall time: 18.8 s


array([[ 191, 3605, 2685,  902,  822],
       [2297, 2747, 4337, 2134, 1170],
       [2747, 3377,  557, 2297,  527],
       ...,
       [4337, 2297, 3575, 2134, 3117],
       [ 655, 2747, 2297,  298, 3695],
       [ 557, 1317, 3679, 4054, 2447]], dtype=int32)

In [21]:
recommendations.shape

(2500, 5)

In [22]:
def reduce_dims(df, dims=2, method='pca'):
    
    assert method in ['pca', 'tsne'], 'Неверно указан метод'
    
    if method=='pca':
        pca = PCA(n_components=dims)
        components = pca.fit_transform(df)
    elif method == 'tsne':
        tsne = TSNE(n_components=dims, learning_rate=250, random_state=42)
        components = tsne.fit_transform(df)
    else:
        print('Error')
        
    colnames = ['component_' + str(i) for i in range(1, dims+1)]
    return pd.DataFrame(data = components, columns = colnames) 


def display_components_in_2D_space(components_df, labels='category', marker='D'):
    
    groups = components_df.groupby(labels)

    # Plot
    fig, ax = plt.subplots(figsize=(12,8))
    ax.margins(0.05) # Optional, just adds 5% padding to the autoscaling
    for name, group in groups:
        ax.plot(group.component_1, group.component_2, 
                marker='o', ms=6,
                linestyle='',
                alpha=0.7,
                label=name)
    ax.legend(loc='center left', bbox_to_anchor=(1.02, 0.5))

    plt.xlabel('component_1')
    plt.ylabel('component_2') 
    plt.show()

In [23]:
model.item_factors.shape

(5001, 64)

In [24]:
category = []

for idx in range(model.item_factors.shape[0]):

    try:
        cat = item_features.loc[item_features['item_id'] == id_to_itemid[idx], 'department'].values[0]
        category.append(cat)
    except:
        category.append('UNKNOWN')

In [25]:
print(category[:10])

['MISC SALES TRAN', 'KIOSK-GAS', 'MISC SALES TRAN', 'KIOSK-GAS', 'MISC SALES TRAN', 'MISC SALES TRAN', 'KIOSK-GAS', 'GROCERY', 'GROCERY', 'GROCERY']


In [26]:
%%time
item_emb_tsne = reduce_dims(model.item_factors, dims=2, method='tsne') # 5001 х 100  ---> 5001 x 2
item_emb_tsne['category'] = category  # Добавляем категорию
item_emb_tsne = item_emb_tsne[item_emb_tsne['category'] != 'UNKNOWN']

Wall time: 21.1 s


In [27]:
item_emb_tsne.head(2)

component_1  component_2         category
0     0.791423    -1.774498  MISC SALES TRAN
1     9.989024   -13.746145        KIOSK-GAS

In [28]:
item_1 = model.item_factors[1]
item_2 = model.item_factors[2]

In [29]:
item_1.shape

(64,)

In [30]:
example_item_row_id = 3606

In [31]:
id_to_itemid[example_item_row_id]

1101010

In [32]:
closest_items = [ id_to_itemid[row_id] for row_id, score in model.similar_items(example_item_row_id, N=5)]

In [33]:
item_features[item_features.item_id.isin(closest_items)]

item_id  manufacturer department     brand commodity_desc  \
5915    819978          3544       DELI  National        CHEESES   
18923   933835          3705       DELI  National     DELI MEATS   
28449  1018740          3516       DELI  National     DELI MEATS   
31252  1043751          4376       DELI  National     DELI MEATS   
37665  1101010          3862       DELI  National     DELI MEATS   

           sub_commodity_desc curr_size_of_product  
5915   CHEESE: PROCESSED BULK                       
18923     MEAT: SAUS DRY BULK                       
28449           MEAT:HAM BULK                       
31252           MEAT:HAM BULK                       
37665    MEAT: LUNCHMEAT BULK

In [34]:
model.similar_users(userid_to_id[10], N=5)

[(9, 1.0),
 (790, 0.9808803),
 (2149, 0.97438735),
 (1788, 0.9727595),
 (354, 0.9722667)]

##### Теперь поэксперементируем

In [109]:
summary_data = pd.DataFrame(data=[[0, 0, 0, 0]], 
                            columns=['factors', 'regularization', 'iterations',  'precision'])

cnt = 0
factors = [50, 100, 150]
regularization = [0.01, 0.001, 0.0001]
iterations = [15, 50, 100]

for fact in factors:
    for reg in regularization:
        for iterat in iterations:
            name = f"ALS_{fact}_{reg}_{iterat}"         
            model = AlternatingLeastSquares(factors=fact, 
                                            regularization=reg,
                                            iterations=iterat, 
                                            calculate_training_loss=True, 
                                            num_threads=10,
                                            use_gpu=False)

            model.fit(csr_matrix(user_item_matrix).T.tocsr(),  # На вход item-user matrix
                      show_progress=False)

            recs = model.recommend(userid=userid_to_id[2],  # userid - id от 0 до N
                                   user_items=csr_matrix(user_item_matrix).tocsr(),   # на вход user-item matrix
                                   N=5, # кол-во рекомендаций 
                                   filter_already_liked_items=False, 
                                   filter_items=None, 
                                   recalculate_user=True)
    
            result['als'] = result['user_id'].apply(lambda x: get_recommendations(x, model=model, N=5))
            score = result.apply(lambda row: precision_at_k(row['als'], row['actual']), axis=1).mean()
            # print(f"For factors {fact}, regularization {reg}, iterations {iterat} Precicion = {score}")
            
            summary_data.loc[cnt, ['factors', 'regularization', 'iterations', 'precision']] = [fact, reg, iterat, score]

            cnt += 1

In [110]:
summary_data

factors  regularization  iterations  precision
0      50.0          0.0100        15.0   0.160221
1      50.0          0.0100        50.0   0.155299
2      50.0          0.0100       100.0   0.154495
3      50.0          0.0010        15.0   0.155399
4      50.0          0.0010        50.0   0.153089
5      50.0          0.0010       100.0   0.152486
6      50.0          0.0001        15.0   0.152587
7      50.0          0.0001        50.0   0.155701
8      50.0          0.0001       100.0   0.153089
9     100.0          0.0100        15.0   0.149573
10    100.0          0.0100        50.0   0.145655
11    100.0          0.0100       100.0   0.146359
12    100.0          0.0010        15.0   0.149674
13    100.0          0.0010        50.0   0.143345
14    100.0          0.0010       100.0   0.144751
15    100.0          0.0001        15.0   0.147162
16    100.0          0.0001        50.0   0.144550
17    100.0          0.0001       100.0   0.144450
18    150.0          0.0100        15.0   0.143044
19    150.0          0.0100        50.0   0.136213
20    150.0          0.0100       100.0   0.132295
21    150.0          0.0010        15.0   0.139427
22    150.0          0.0010        50.0   0.134204
23    150.0          0.0010       100.0   0.130085
24    150.0          0.0001        15.0   0.138925
25    150.0          0.0001        50.0   0.133400
26    150.0          0.0001       100.0   0.132396

In [119]:
summary_data.sort_values('precision', ascending=False)

factors  regularization  iterations  precision
0      50.0          0.0100        15.0   0.160221
7      50.0          0.0001        50.0   0.155701
3      50.0          0.0010        15.0   0.155399
1      50.0          0.0100        50.0   0.155299
2      50.0          0.0100       100.0   0.154495
4      50.0          0.0010        50.0   0.153089
8      50.0          0.0001       100.0   0.153089
6      50.0          0.0001        15.0   0.152587
5      50.0          0.0010       100.0   0.152486
12    100.0          0.0010        15.0   0.149674
9     100.0          0.0100        15.0   0.149573
15    100.0          0.0001        15.0   0.147162
11    100.0          0.0100       100.0   0.146359
10    100.0          0.0100        50.0   0.145655
14    100.0          0.0010       100.0   0.144751
16    100.0          0.0001        50.0   0.144550
17    100.0          0.0001       100.0   0.144450
13    100.0          0.0010        50.0   0.143345
18    150.0          0.0100        15.0   0.143044
21    150.0          0.0010        15.0   0.139427
24    150.0          0.0001        15.0   0.138925
19    150.0          0.0100        50.0   0.136213
22    150.0          0.0010        50.0   0.134204
25    150.0          0.0001        50.0   0.133400
26    150.0          0.0001       100.0   0.132396
20    150.0          0.0100       100.0   0.132295
23    150.0          0.0010       100.0   0.130085

##### Вывод: 
Наилучший результат получился при следующих пкараметрах factors = 50,	regularization = 0.01, iterations = 15 и составил - 0.160221, что выше чем на вебинаре (0.1574083375188326)